File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/repositories/HouseholdMemberRepository.java
package com.decrux.pocketr.api.repositories;

import com.decrux.pocketr.api.entities.db.household.HouseholdMember;
import com.decrux.pocketr.api.entities.db.household.HouseholdMemberId;
import com.decrux.pocketr.api.entities.db.household.MemberStatus;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
public interface HouseholdMemberRepository extends JpaRepository<HouseholdMember, HouseholdMemberId> {

    List<HouseholdMember> findByUserUserId(long userId);

    List<HouseholdMember> findByUserUserIdAndStatus(long userId, MemberStatus status);

    boolean existsByUserUserIdAndStatus(long userId, MemberStatus status);

    boolean existsByUserUserIdAndStatusAndHouseholdIdNot(long userId, MemberStatus status, UUID householdId);

    HouseholdMember findByHouseholdIdAndUserUserId(UUID householdId, long userId);

    default Optional<HouseholdMember> findOptionalByHouseholdIdAndUserUserId(UUID householdId, long userId) {
        return Optional.ofNullable(findByHouseholdIdAndUserUserId(householdId, userId));
    }

    List<HouseholdMember> findByHouseholdIdAndStatus(UUID householdId, MemberStatus status);

    List<HouseholdMember> findByHouseholdId(UUID householdId);
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr/api/repositories/HouseholdMemberRepository.kt
package com.decrux.pocketr.api.repositories

import com.decrux.pocketr.api.entities.db.household.HouseholdMember
import com.decrux.pocketr.api.entities.db.household.HouseholdMemberId
import com.decrux.pocketr.api.entities.db.household.MemberStatus
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import java.util.UUID

@Repository
interface HouseholdMemberRepository : JpaRepository<HouseholdMember, HouseholdMemberId> {
    fun findByUserUserId(userId: Long): List<HouseholdMember>

    fun findByUserUserIdAndStatus(
        userId: Long,
        status: MemberStatus,
    ): List<HouseholdMember>

    fun existsByUserUserIdAndStatus(
        userId: Long,
        status: MemberStatus,
    ): Boolean

    fun existsByUserUserIdAndStatusAndHouseholdIdNot(
        userId: Long,
        status: MemberStatus,
        householdId: UUID,
    ): Boolean

    fun findByHouseholdIdAndUserUserId(
        householdId: UUID,
        userId: Long,
    ): HouseholdMember?

    fun findByHouseholdIdAndStatus(
        householdId: UUID,
        status: MemberStatus,
    ): List<HouseholdMember>

    fun findByHouseholdId(householdId: UUID): List<HouseholdMember>
}
